File: lib/diff-analyzer.js
/**
 * Diff Analyzer Agent
 * Analyzes git diff output to extract meaningful information about changes
 */

const { execSync } = require('child_process');
const path = require('path');

class DiffAnalyzer {
    constructor() {
        this.fileExtensions = {
            '.py': 'python',
            '.js': 'javascript',
            '.ts': 'typescript',
            '.jsx': 'react',
            '.tsx': 'react',
            '.go': 'golang',
            '.java': 'java',
            '.cpp': 'cpp',
            '.c': 'c',
            '.rs': 'rust',
            '.rb': 'ruby',
            '.php': 'php',
            '.css': 'css',
            '.scss': 'sass',
            '.html': 'html',
            '.md': 'documentation',
            '.json': 'config',
            '.yaml': 'config',
            '.yml': 'config',
            '.xml': 'config',
            '.toml': 'config',
            '.dockerfile': 'docker',
            '.sql': 'database',
            '.sh': 'script',
            '.bash': 'script',
            '.zsh': 'script'
        };

        this.changePatterns = {
            'feature': [
                /add.*function/i, /new.*class/i, /implement.*/i,
                /create.*/i, /introduce.*/i, /build.*/i
            ],
            'fix': [
                /fix.*bug/i, /resolve.*/i, /correct.*/i,
                /patch.*/i, /repair.*/i, /address.*issue/i
            ],
            'refactor': [
                /refactor.*/i, /restructure.*/i, /reorganize.*/i,
                /clean.*up/i, /simplify.*/i, /optimize.*/i
            ],
            'docs': [
                /update.*documentation/i, /add.*comment/i, /improve.*readme/i,
                /document.*/i, /add.*docstring/i
            ],
            'style': [
                /format.*/i, /indent.*/i, /whitespace.*/i,
                /style.*/i, /lint.*/i, /prettier.*/i
            ],
            'test': [
                /add.*test/i, /test.*/i, /spec.*/i, /mock.*/i
            ],
            'chore': [
                /update.*dependency/i, /upgrade.*/i, /maintenance.*/i,
                /cleanup.*/i, /remove.*unused/i
            ]
        };
    }

    /**
     * Get current git status information
     */
    getGitStatus() {
        try {
            const result = execSync('git status --porcelain', { encoding: 'utf8' });
            
            const status = {
                added: [],
                modified: [],
                deleted: [],
                renamed: []
            };

            const lines = result.trim().split('\n');
            for (const line of lines) {
                if (!line) continue;

                const statusCode = line.substring(0, 2);
                const filename = line.substring(3);

                if (statusCode.includes('A')) {
                    status.added.push(filename);
                } else if (statusCode.includes('M')) {
                    status.modified.push(filename);
                } else if (statusCode.includes('D')) {
                    status.deleted.push(filename);
                } else if (statusCode.includes('R')) {
                    status.renamed.push(filename);
                }
            }

            return status;
        } catch (error) {
            return { added: [], modified: [], deleted: [], renamed: [] };
        }
    }

    /**
     * Get diff statistics (lines added/removed)
     */
    getDiffStats() {
        try {
            const result = execSync('git diff --cached --numstat', { encoding: 'utf8' });
            
            let added = 0, removed = 0, filesChanged = 0;

            const lines = result.trim().split('\n');
            for (const line of lines) {
                if (!line) continue;
                
                const parts = line.split('\t');
                if (parts.length >= 2) {
                    try {
                        added += parts[0] !== '-' ? parseInt(parts[0]) : 0;
                        removed += parts[1] !== '-' ? parseInt(parts[1]) : 0;
                        filesChanged++;
                    } catch (error) {
                        continue;
                    }
                }
            }

            return {
                files_changed: filesChanged,
                insertions: added,
                deletions: removed
            };
        } catch (error) {
            return { files_changed: 0, insertions: 0, deletions: 0 };
        }
    }

    /**
     * Get the staged diff content
     */
    getStagedDiff() {
        try {
            const result = execSync('git diff --cached', { encoding: 'utf8' });
            return result;
        } catch (error) {
            return '';
        }
    }

    /**
     * Analyze the types of files being changed
     */
    analyzeFileTypes(files) {
        const fileTypes = {};

        for (const file of files) {
            const ext = path.extname(file).toLowerCase();
            const fileType = this.fileExtensions[ext] || 'other';
            fileTypes[fileType] = (fileTypes[fileType] || 0) + 1;
        }

        return fileTypes;
    }

    /**
     * Detect the type of change based on diff content
     */
    detectChangeType(diffContent) {
        const diffLower = diffContent.toLowerCase();

        // Count matches for each change type
        const typeScores = {};
        for (const [changeType, patterns] of Object.entries(this.changePatterns)) {
            let score = 0;
            for (const pattern of patterns) {
                const matches = (diffLower.match(pattern) || []).length;
                score += matches;
            }
            typeScores[changeType] = score;
        }

        // Return the type with the highest score, default to 'feat' if no matches
        if (!typeScores || Math.max(...Object.values(typeScores)) === 0) {
            return 'feat';
        }

        return Object.keys(typeScores).reduce((a, b) => 
            typeScores[a] > typeScores[b] ? a : b
        );
    }

    /**
     * Extract function and class names from diff
     */
    extractFunctionsAndClasses(diffContent) {
        const patterns = [
            /\+.*def\s+(\w+)/g,        // Python functions
            /\+.*function\s+(\w+)/g,   // JavaScript functions
            /\+.*class\s+(\w+)/g,      // Class definitions
            /\+.*const\s+(\w+)\s*=/g,  // Const definitions
            /\+.*let\s+(\w+)\s*=/g,    // Let definitions
            /\+.*var\s+(\w+)\s*=/g,    // Var definitions
        ];

        const entities = [];
        for (const pattern of patterns) {
            let match;
            while ((match = pattern.exec(diffContent)) !== null) {
                entities.push(match[1]);
            }
        }

        // Remove duplicates
        return [...new Set(entities)];
    }

    /**
     * Perform comprehensive analysis of staged changes
     */
    analyzeChanges() {
        const status = this.getGitStatus();
        const diffStats = this.getDiffStats();
        const diffContent = this.getStagedDiff();

        // Get all changed files
        const allFiles = [];
        for (const fileList of Object.values(status)) {
            allFiles.push(...fileList);
        }

        const fileTypes = this.analyzeFileTypes(allFiles);
        const changeType = this.detectChangeType(diffContent);
        const entities = this.extractFunctionsAndClasses(diffContent);

        // Determine scope based on file types
        const scope = this.determineScope(fileTypes, allFiles);

        return {
            status,
            stats: diffStats,
            file_types: fileTypes,
            change_type: changeType,
            scope,
            entities: entities.slice(0, 5), // Limit to first 5 entities
            diff_content: diffContent
        };
    }

    /**
     * Determine the scope of changes
     */
    determineScope(fileTypes, files) {
        if (Object.keys(fileTypes).length === 1) {
            return Object.keys(fileTypes)[0];
        }

        // Check for specific patterns
        if (files.some(f => f.toLowerCase().includes('test'))) {
            return 'test';
        }
        if (files.some(f => f.toLowerCase().includes('doc') || f.toLowerCase().includes('readme'))) {
            return 'docs';
        }
        if (files.some(f => f.toLowerCase().includes('config') || f.includes('.json') || f.includes('.yaml'))) {
            return 'config';
        }

        // Return the most common file type
        if (Object.keys(fileTypes).length > 0) {
            return Object.keys(fileTypes).reduce((a, b) => 
                fileTypes[a] > fileTypes[b] ? a : b
            );
        }

        return null;
    }
}

module.exports = DiffAnalyzer;

File: lib/diff_analyzer.py
#!/usr/bin/env python3
"""
Diff Analyzer Agent
Analyzes git diff output to extract meaningful information about changes
"""

import re
import subprocess
import json
from typing import Dict, List, Tuple, Optional
from pathlib import Path

class DiffAnalyzer:
    def __init__(self):
        self.file_extensions = {
            '.py': 'python',
            '.js': 'javascript',
            '.ts': 'typescript',
            '.jsx': 'react',
            '.tsx': 'react',
            '.go': 'golang',
            '.java': 'java',
            '.cpp': 'cpp',
            '.c': 'c',
            '.rs': 'rust',
            '.rb': 'ruby',
            '.php': 'php',
            '.css': 'css',
            '.scss': 'sass',
            '.html': 'html',
            '.md': 'documentation',
            '.json': 'config',
            '.yaml': 'config',
            '.yml': 'config',
            '.xml': 'config',
            '.toml': 'config',
            '.dockerfile': 'docker',
            '.sql': 'database',
            '.sh': 'script',
            '.bash': 'script',
            '.zsh': 'script'
        }
        
        self.change_patterns = {
            'feature': [
                r'add.*function', r'new.*class', r'implement.*', 
                r'create.*', r'introduce.*', r'build.*'
            ],
            'fix': [
                r'fix.*bug', r'resolve.*', r'correct.*',
                r'patch.*', r'repair.*', r'address.*issue'
            ],
            'refactor': [
                r'refactor.*', r'restructure.*', r'reorganize.*',
                r'clean.*up', r'simplify.*', r'optimize.*'
            ],
            'docs': [
                r'update.*documentation', r'add.*comment', r'improve.*readme',
                r'document.*', r'add.*docstring'
            ],
            'style': [
                r'format.*', r'indent.*', r'whitespace.*',
                r'style.*', r'lint.*', r'prettier.*'
            ],
            'test': [
                r'add.*test', r'test.*', r'spec.*', r'mock.*'
            ],
            'chore': [
                r'update.*dependency', r'upgrade.*', r'maintenance.*',
                r'cleanup.*', r'remove.*unused'
            ]
        }

    def get_git_status(self) -> Dict[str, List[str]]:
        """Get current git status information"""
        try:
            result = subprocess.run(['git', 'status', '--porcelain'], 
                                  capture_output=True, text=True, check=True)
            
            status = {'added': [], 'modified': [], 'deleted': [], 'renamed': []}
            
            for line in result.stdout.strip().split('\n'):
                if not line:
                    continue
                    
                status_code = line[:2]
                filename = line[3:]
                
                if 'A' in status_code:
                    status['added'].append(filename)
                elif 'M' in status_code:
                    status['modified'].append(filename)
                elif 'D' in status_code:
                    status['deleted'].append(filename)
                elif 'R' in status_code:
                    status['renamed'].append(filename)
                    
            return status
        except subprocess.CalledProcessError:
            return {'added': [], 'modified': [], 'deleted': [], 'renamed': []}

    def get_diff_stats(self) -> Dict[str, int]:
        """Get diff statistics (lines added/removed)"""
        try:
            result = subprocess.run(['git', 'diff', '--cached', '--numstat'], 
                                  capture_output=True, text=True, check=True)
            
            added, removed = 0, 0
            files_changed = 0
            
            for line in result.stdout.strip().split('\n'):
                if not line:
                    continue
                parts = line.split('\t')
                if len(parts) >= 2:
                    try:
                        added += int(parts[0]) if parts[0] != '-' else 0
                        removed += int(parts[1]) if parts[1] != '-' else 0
                        files_changed += 1
                    except ValueError:
                        continue
                        
            return {
                'files_changed': files_changed,
                'insertions': added,
                'deletions': removed
            }
        except subprocess.CalledProcessError:
            return {'files_changed': 0, 'insertions': 0, 'deletions': 0}

    def get_staged_diff(self) -> str:
        """Get the staged diff content"""
        try:
            result = subprocess.run(['git', 'diff', '--cached'], 
                                  capture_output=True, text=True, check=True)
            return result.stdout
        except subprocess.CalledProcessError:
            return ""

    def analyze_file_types(self, files: List[str]) -> Dict[str, int]:
        """Analyze the types of files being changed"""
        file_types = {}
        
        for file in files:
            ext = Path(file).suffix.lower()
            file_type = self.file_extensions.get(ext, 'other')
            file_types[file_type] = file_types.get(file_type, 0) + 1
            
        return file_types

    def detect_change_type(self, diff_content: str) -> str:
        """Detect the type of change based on diff content"""
        diff_lower = diff_content.lower()
        
        # Count matches for each change type
        type_scores = {}
        for change_type, patterns in self.change_patterns.items():
            score = 0
            for pattern in patterns:
                matches = len(re.findall(pattern, diff_lower))
                score += matches
            type_scores[change_type] = score
        
        # Return the type with the highest score, default to 'feat' if no matches
        if not type_scores or max(type_scores.values()) == 0:
            return 'feat'
        
        return max(type_scores, key=type_scores.get)

    def extract_functions_and_classes(self, diff_content: str) -> List[str]:
        """Extract function and class names from diff"""
        patterns = [
            r'\+.*def\s+(\w+)',  # Python functions
            r'\+.*function\s+(\w+)',  # JavaScript functions
            r'\+.*class\s+(\w+)',  # Class definitions
            r'\+.*const\s+(\w+)\s*=',  # Const definitions
            r'\+.*let\s+(\w+)\s*=',  # Let definitions
            r'\+.*var\s+(\w+)\s*=',  # Var definitions
        ]
        
        entities = []
        for pattern in patterns:
            matches = re.findall(pattern, diff_content, re.MULTILINE)
            entities.extend(matches)
        
        return list(set(entities))  # Remove duplicates

    def analyze_changes(self) -> Dict:
        """Perform comprehensive analysis of staged changes"""
        status = self.get_git_status()
        diff_stats = self.get_diff_stats()
        diff_content = self.get_staged_diff()
        
        # Get all changed files
        all_files = []
        for file_list in status.values():
            all_files.extend(file_list)
        
        file_types = self.analyze_file_types(all_files)
        change_type = self.detect_change_type(diff_content)
        entities = self.extract_functions_and_classes(diff_content)
        
        # Determine scope based on file types
        scope = self.determine_scope(file_types, all_files)
        
        return {
            'status': status,
            'stats': diff_stats,
            'file_types': file_types,
            'change_type': change_type,
            'scope': scope,
            'entities': entities[:5],  # Limit to first 5 entities
            'diff_content': diff_content
        }

    def determine_scope(self, file_types: Dict[str, int], files: List[str]) -> Optional[str]:
        """Determine the scope of changes"""
        if len(file_types) == 1:
            return list(file_types.keys())[0]
        
        # Check for specific patterns
        if any('test' in f.lower() for f in files):
            return 'test'
        if any('doc' in f.lower() or 'readme' in f.lower() for f in files):
            return 'docs'
        if any('config' in f.lower() or '.json' in f or '.yaml' in f for f in files):
            return 'config'
        
        # Return the most common file type
        if file_types:
            return max(file_types, key=file_types.get)
        
        return None

if __name__ == "__main__":
    analyzer = DiffAnalyzer()
    analysis = analyzer.analyze_changes()
    print(json.dumps(analysis, indent=2))